gpflow/training/monitor.py

```diff
@@ -631,16 +632,27 @@ class BaseTensorBoardTask(MonitorTask):
     providing these values as the input values dictionary.
     """
 
+    _all_file_writers = {}  # type: Dict[str, tf.summary.FileWriter]
```

Comment:
Can we avoid to use global variable here? Most of the time this will not be an issue, but nevertheless this dictionary can grow fast. Simple example: using loop and create different directories for tensorboard, like 'path_1', 'path_2', ... 'path_n', and, btw,  they will never be closed - that's another issue. Options:
* pass writer explicitly (it is easy to do) and create one only when 'str' passed.
* create abstraction for TensorBoard and pass this one instead of the file writer.

Reply:
I agree that using a global variable is not nice in general and keeping not closed file writers is bad (although in practice it's very unlikely cause problems). So I made some improvements.
I didn't want to pass it over to a user to deal with the FileWriter for two reasons. One would expect the framework to do things like opening and closing files. More importantly, a user is likely to run into the same problem as us - trying to open multiple writers (not everyone reads every line of documentation). TF kindly doesn't raise any exception, it just silently doesn't do anything. So the user will jump to immediate conclusion that the Monitor doesn't work.
Handling writers at the monitor level is also problematic because the user can create more than  one monitor.
So I left the FileWriter cache in a global variable but now I make sure that all writers are closed when the monitor stops running. Unless they are used by another monitor, in which case the last one will close all remaining writers.
I also made the monitor running in a context which slightly simplifies the syntax and ensures that the tasks close their writers.